### PR TITLE
Accept onboard computer heartbeats as telemetry

### PR DESCRIPTION
**Pull request by okalachev**
This PR solves #10042 problem, that `STATUSTEXT` messages are not passed, when there is no GCS.
Also, this is a partial implementation of #7985.

I pass `telemetry_status`, when OBC heartbeats received, but I changed the commander logic, for not considering OBC as a datalink.

So `NAV_DLL_ACT`, `COM_DL_LOSS_T`, `COM_DL_REG_T` are only about GCS link, not OBC (not renaming for backwards compatibility).

The OBC failsafe params, like `NAV_OBC_DLL_ACT`, `COM_OBC_DL_LOSS_T`, `COM_OBC_DL_REG_T` can be added later.

@LorenzMeier , @TSC21 .

### CODE REVIEW

**Comment by TSC21** [comment]
This looks reasonable and can be extended for an OBC datalink failsafe after. Can you please rebase?

**Comment by TSC21** [comment]
@okalachev: http://ci.px4.io:8080/blue/organizations/jenkins/PX4%2FFirmware/detail/PR-10194/2/pipeline#step-205-log-1111

File: src/modules/mavlink/mavlink_receiver.cpp [hunk 1]
```diff
@@ -1854,16 +1854,16 @@ MavlinkReceiver::handle_message_heartbeat(mavlink_message_t *msg)
 		mavlink_heartbeat_t hb;
 		mavlink_msg_heartbeat_decode(msg, &hb);
 
-		/* ignore own heartbeats, accept only heartbeats from GCS */
-		if (msg->sysid != mavlink_system.sysid && hb.type == MAV_TYPE_GCS) {
-
+		/* accept only heartbeats from GCS or Onboard computers */
+		if (hb.type == MAV_TYPE_GCS || hb.type == MAV_TYPE_ONBOARD_CONTROLLER) {
```
Comment (dagar) [inline src/modules/mavlink/mavlink_receiver.cpp]: Is there any reason we can't drop this check entirely now?

File: src/modules/commander/state_machine_helper.cpp [hunk 2]
```diff
@@ -56,7 +56,7 @@ static constexpr const char reason_no_offboard[] = "no offboard";
 static constexpr const char reason_no_rc_and_no_offboard[] = "no RC and no offboard";
 static constexpr const char reason_no_local_position[] = "no local position";
 static constexpr const char reason_no_global_position[] = "no global position";
-static constexpr const char reason_no_datalink[] = "no datalink";
+static constexpr const char reason_no_datalink[] = "no GCS datalink";
```
Comment (dagar) [inline src/modules/commander/state_machine_helper.cpp]: How about ground control instead of GCS?

**Comment by dagar** [comment]
A few minor comments, but overall looks good.

File: src/modules/mavlink/mavlink_receiver.cpp (hunk 1 above)
Comment (okalachev) [reply src/modules/mavlink/mavlink_receiver.cpp]: What if there is a heartbeat from another drone (in a swarm network), charging station, or any other device? Should it be considered as a telemetry link?
I don't think so.

File: src/modules/commander/state_machine_helper.cpp (hunk 2 above)
Comment (okalachev) [reply src/modules/commander/state_machine_helper.cpp]: I don't see a huge difference, but I can change it.

**Comment by okalachev** [comment]
@TSC21 , I found a bug.
The problem starts, when multiple telemetry "sources" are on the same Mavlink instance, so this is the same `telemetry_status` instance, and the same "telemetry" in commander's telemetries array.

So, I don's see an easy solution yet, look like a lot of telemetry logic should be changed.

**Comment by TSC21** [comment]
@okalachev that probably means we need a different `telemetry_status` topic instance for each of the telemetry sources. Probably multitopic is required.

**Comment by okalachev** [comment]
I guess the maximum number of topic instances are `ORB_MULTI_MAX_INSTANCES`, which is 4. But the number of telemetry "sources" that way would be something like 255 (systems) * 255 (systems' components).

So, maybe not to use multitopic for that. Instead, to add system and component ids to `telemetry_status`?

**Comment by TSC21** [comment]
> I guess the maximum number of topic instances are `ORB_MULTI_MAX_INSTANCES`, which is 4. But the number of telemetry "sources" that way would be something like 255 (systems) * 255 (systems' components).

Well I was thinking on limit the telemetry sources based on type and not on sys+comp id. But I guess that would make more sense yes. Can you propose that in a new issue as a RFC? Probably better to have a discussion about it first before one commits to it.

**Comment by okalachev** [comment]
You mean to write an RFC for adding sys and comp ids to `telemetry_status`?

I'm not sure this is so large change to make an RFC.

**Comment by TSC21** [comment]
> I'm not sure this is so large change to make an RFC.

It's a conceptual/structural change/feature that needs to be discussed.

**Comment by LorenzMeier** [comment]
Could you please rebase? Thanks!

**Comment by okalachev** [comment]
Unfortunately, it doesn't make sense for now, because this change introduces a new bug, that I described in previous messages.

I don't see an easy solution yet, without rewriting `telemetry_status` logic.

**Comment by TSC21** [comment]
@okalachev what's the status of this? Have you thought of a way of bringing this into a state that does work?

**Comment by okalachev** [comment]
@TSC21, sorry, unfortunately I haven't worked at this yet. The easy workaround is to send fake GCS heartbeats from ROS code. The conversation can be continued at https://github.com/PX4/Firmware/issues/10042.

This PR can be closed I guess, as the implementation is incorrect.

**Comment by dagar** [comment]
Revisiting this idea - https://github.com/PX4/Firmware/pull/10933

**Pull request closed without merging**